### PR TITLE
Announce forthcoming requirement of Java 11

### PR DESCRIPTION
**Pull request by basil**
Warn users who run Jenkins with a deprecated Java version of the forthcoming Java 8 EOL. The date has not yet been finalized, hence this PR being in draft state. For the sake of example, I have chosen [Flag Day](https://en.wikipedia.org/wiki/Flag_day_(computing)) 2022. Once stakeholders like @MarkEWaite have made the decision, this PR can be updated with the final date.

### Proposed changelog entries

Add an administrative monitor announcing the forthcoming requirement of Java 11.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by daniel-beck** [comment]
The problem with re-using an admin monitor is that the original one may have been dismissed already. That's why I wanted two monitors for #5337, to have the second one show up again once the admin sets up a cloud or agents, even if the first one had been dismissed before.

**Comment by basil** [comment]
OK, so should we automatically "un-dismiss" the original dismissal to display this new warning, or just copy-paste the existing monitor into a new one?

**Comment by daniel-beck** [comment]
> OK, so should we automatically "un-dismiss" the original dismissal to display this new warning, or just copy-paste the existing monitor into a new one?

IMO the latter, but perhaps we can minimize the duplication by subclassing or similar?

File: core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.properties
Comment (timja) [inline core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.properties:1]: this date seems a bit far away? shouldn't it be after the next LTS baseline (9th March).

not 100% sure how to express but my thoughts would be that line will continue under Java 8 but the next weekly would drop support?

**Review by timja — COMMENTED** [review]
nice, updating the ID should be enough imo to force it to be re-triggered

**Comment by basil** [comment]
> or perhaps just an ID changed?

Done

File: core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.properties
Comment (basil) [reply core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.properties:1]: > this date seems a bit far away? shouldn't it be after the next LTS baseline (9th March).

Please refer to the PR description:

> The date has not yet been finalized, hence this PR being in draft state. For the sake of example, I have chosen [Flag Day](https://en.wikipedia.org/wiki/Flag_day_(computing)) 2022. Once stakeholders like @MarkEWaite have made the decision, this PR can be updated with the final date.

> not 100% sure how to express but my thoughts would be that line will continue under Java 8 but the next weekly would drop support?

Different branches can have different dates.

File: core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.jelly
Comment (timja) [inline core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.jelly:28]: slight over engineering I think, the monitor only works with java 8 but I guess no harm and could be useful later

**Review by timja — APPROVED** [review]
lgtm pending date

File: core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.jelly
Comment (basil) [reply core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.jelly:28]: Good communication about migrations ought to include at least two pieces of information, I think: what you're currently using, and what you should be using. Yes the monitor only works with Java 8, but including the Java version in the message ought to help users understand more clearly what needs to change on their end.

Comment (timja) [reply core/src/main/resources/jenkins/monitor/JavaVersionRecommendationAdminMonitor/message.jelly:28]: I get that, hardcoding java 8 would work too though 😄. no issue with this though just was a bit confused it was dynamically getting something that is always going to be 8 (with the current code)

File: core/src/main/java/jenkins/monitor/JavaVersionRecommendationAdminMonitor.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/monitor/JavaVersionRecommendationAdminMonitor.java:50]:
```suggestion
        super(JavaVersionRecommendationAdminMonitor.class.getName() + "-2");
```

Assuming this is legal, to not make people look for a class named `JavaVersionRecommendationAdminMonitor2`?

Comment (basil) [reply core/src/main/java/jenkins/monitor/JavaVersionRecommendationAdminMonitor.java]: Done

**Comment by basil** [comment]
@MarkEWaite When you have a date, can you please open a suggestion against this PR with the final date? Then I will mark this PR as ready for review.

**Comment by jglick** [comment]
see #6083

**Comment by basil** [comment]
Has the feedback from [this review](https://github.com/jenkinsci/jenkins/pull/6092#pullrequestreview-837618132) been addressed? If so, is there a reason this PR has not yet been approved?

**Comment by timja** [comment]
> Has the feedback from [this review](https://github.com/jenkinsci/jenkins/pull/6092#pullrequestreview-837618132) been addressed? If so, is there a reason this PR has not yet been approved?

Just waiting on a date which will probably be the weekly one not the LTS one good apart from that

**Comment by basil** [comment]
> Just waiting on a date

I was addressing Daniel, but I explicitly noted in the PR description that deciding on the date is explicitly out of scope for this PR. (I saw that you started a separate mailing list thread about the date, which makes sense.) So I see no reason why this PR could not be approved pending a final decision regarding the date.

**Review by timja — APPROVED** [review]
Fine to me pending date

**Comment by jglick** [comment]
886db02798be4bd573bd719f62e6de8db6ae1980 reference: https://groups.google.com/g/jenkinsci-dev/c/YghQ0YP4m78/m/CMr-FD49EAAJ

**Comment by basil** [comment]
Requesting a review from @MarkEWaite yet again. I have responded to every piece of feedback that has been left so far, including conceding to a September date, so I am not sure what action items remain for approval of this PR. I note that Mark did approve jenkinsci/packaging#306, which is along the lines of this PR but more vague (i.e., "in the future" rather than "in September"). I could possibly go in that direction, but I do not like vagueness and I will not do so unless requested to do so.

Mark, could you please indicate clearly what action items remain from your perspective in order for this PR to be approved? I am happy to discuss this in whatever forum you deem appropriate: core pull request, JEP pull request, developer mailing list, community forum, SIG meeting, personal communication, etc. If there is a specific action item to be taken on my side, I am happy to take it. An explicit response will help me understand where we are in the process and where to focus going forward.

**Review by MarkEWaite — APPROVED** [review]
No action items needed from my perspective.  I think we're ready to declare that the September 2022 LTS release will require Java 11 or newer.

**Review by MarkEWaite — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Comment by basil** [comment]
I added a proposed changelog entry reading "Add an administrative monitor announcing the forthcoming Java 8 EOL." Happy to adjust this based on feedback.

**Pull request merged**
